common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesAccountCredentialCache.java

```diff
@@ -103,21 +103,37 @@ public SharedPreferencesAccountCredentialCache(
     }
 
     @Override
-    public synchronized void saveAccount(@NonNull final AccountRecord account) {
+    public synchronized void saveAccount(@NonNull final AccountRecord accountToSave) {
         Logger.verbose(TAG, "Saving Account...");
-        Logger.verbose(TAG, "Account type: [" + account.getClass().getSimpleName() + "]");
-        final String cacheKey = mCacheValueDelegate.generateCacheKey(account);
+        Logger.verbose(TAG, "Account type: [" + accountToSave.getClass().getSimpleName() + "]");
+        final String cacheKey = mCacheValueDelegate.generateCacheKey(accountToSave);
         Logger.verbosePII(TAG, "Generated cache key: [" + cacheKey + "]");
-        final String cacheValue = mCacheValueDelegate.generateCacheValue(account);
+
+        // Perform any necessary field merging on the Account to save...
+        final AccountRecord existingAccount = getAccount(cacheKey);
+
+        if (null != existingAccount) {
+            accountToSave.mergeAdditionalFields(existingAccount);
+        }
+
+        final String cacheValue = mCacheValueDelegate.generateCacheValue(accountToSave);
         mSharedPreferencesFileManager.putString(cacheKey, cacheValue);
     }
 
     @Override
-    public synchronized void saveCredential(@NonNull Credential credential) {
+    public synchronized void saveCredential(@NonNull Credential credentialToSave) {
         Logger.verbose(TAG, "Saving credential...");
-        final String cacheKey = mCacheValueDelegate.generateCacheKey(credential);
+        final String cacheKey = mCacheValueDelegate.generateCacheKey(credentialToSave);
         Logger.verbosePII(TAG, "Generated cache key: [" + cacheKey + "]");
-        final String cacheValue = mCacheValueDelegate.generateCacheValue(credential);
+
+        // Perform any necessary field merging on the Credential to save...
+        final Credential existingCredential = getCredential(cacheKey);
+
+        if (null != existingCredential) {
```

Comment:
Can we have this create a new copy of the credential rather than mutate it?

Reply:
I'd like to decline this suggestion: the cache doesn't know how to construct `Credentials` as they're an abstract type. Unless I'm mistaken, implementing the behavior you describe would requires adding clone(), copy constructors or a fair bit of code to inspect the type and reconstruct the instance of the subclass, copying all the relevant properties

Reply:
Does the cacheValueDelegate thingy care about the actual type of the class?  It doesn't seem to.  The abstract class *could* just return an anonymous AccountCredentialBase that overrides that method.

Reply:
OK, too painful.  We'd end up doing custom things to make this happen.  We talked offline, and I know I've seen this before in XML serialization.  In XML there was a lot of painful DOM manipulation to make these extra properties not just still present, but also stable (and comments, too) for all inheritors of a base class.  I want to say that we avoided this for JSON once upon a time by having the serializers gather all of the extra properties into a map on serialization using basically some GSON TypeAdapter trick in one case, and in the other we just treated all objects as generic key-value storage and only enforced typing on a per-key basis.